### PR TITLE
[LegalizeTypes][RISCV] Use SPLAT_VECTOR_PARTS to legalize splat BUILD_VECTOR

### PR DESCRIPTION
**Pull request by topperc**
If the element type needs to be expanded, we can use SPLAT_VECTOR_PARTS if the target supports it.

There's already a DAGCombine to turn BUILD_VECTOR into SPLAT_VECTOR if the target makes SPLAT_VECTOR legal, but it doesn't fire for vectors that need to be split.

Alternatively, we could tweak the existig DAGCombine to form SPLAT_VECTOR if the type will be split to a legal vector type that supports SPLAT_VECTOR. That requires iterating through getTypeToTransformTo until we reach a legal type.

This is an alternative to #107205

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp [hunk 1]
```diff
@@ -376,6 +376,15 @@ SDValue DAGTypeLegalizer::ExpandOp_BUILD_VECTOR(SDNode *N) {
   assert(OldVT == VecVT.getVectorElementType() &&
          "BUILD_VECTOR operand type doesn't match vector element type!");
 
+  if (VecVT.isInteger() && TLI.isOperationLegal(ISD::SPLAT_VECTOR, VecVT) &&
```
Comment (preames) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp:379]: Why do we need the SPLAT_VECTOR legality?  Isn't SPLAT_VECTOR_PARTS enough?

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-llrint.ll [hunk 2]
```diff
@@ -14,9 +14,11 @@ define <1 x i64> @llrint_v1i64_v1f32(<1 x float> %x) {
 ; RV32-NEXT:    vsetivli zero, 1, e32, m1, ta, ma
 ; RV32-NEXT:    vfmv.f.s fa0, v8
 ; RV32-NEXT:    call llrintf
-; RV32-NEXT:    vsetivli zero, 2, e32, mf2, ta, ma
-; RV32-NEXT:    vmv.v.x v8, a0
-; RV32-NEXT:    vslide1down.vx v8, v8, a1
+; RV32-NEXT:    sw a1, 4(sp)
+; RV32-NEXT:    sw a0, 0(sp)
+; RV32-NEXT:    mv a0, sp
+; RV32-NEXT:    vsetivli zero, 1, e64, m1, ta, ma
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-llrint.ll:20]: Given the results on short VL, we'd probably be better going through the buildvector lowering, or adapting some of that lowering to work on short splat_vector_parts.  Using the strided load may not always be our best fallback.  

(Optional follow up only)

**Review by preames — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp:379]: I think we don't default SPLAT_VECTOR_PARTS to Expand so it started crashing other targets.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-llrint.ll (hunk 2 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-llrint.ll:20]: Agree. The strided load is not good given what we know now about microarchitectures.

**Review by lukel97 — APPROVED** [review]

**Pull request merged**